crates/biome_graphql_parser/src/parser/definitions/operation.rs

```diff
@@ -69,69 +62,213 @@ impl ParseRecovery for SelectionListParseRecovery {
     }
 }
 
-pub(crate) struct SelectionList;
+#[derive(Default)]
+struct VariableDefinitionList;
 
-impl SelectionList {
-    pub(crate) fn new() -> Self {
-        Self
-    }
-}
-
-impl ParseNodeList for SelectionList {
+impl ParseNodeList for VariableDefinitionList {
     type Kind = GraphqlSyntaxKind;
     type Parser<'source> = GraphqlParser<'source>;
 
-    const LIST_KIND: Self::Kind = GRAPHQL_SELECTION_LIST;
+    const LIST_KIND: Self::Kind = GRAPHQL_VARIABLE_DEFINITION_LIST;
 
     fn parse_element(&mut self, p: &mut Self::Parser<'_>) -> ParsedSyntax {
-        parse_selection(p)
+        parse_variable_definition(p)
     }
 
     fn is_at_list_end(&self, p: &mut Self::Parser<'_>) -> bool {
-        p.at(T!['}'])
+        p.at(T![')'])
     }
 
     fn recover(
         &mut self,
         p: &mut Self::Parser<'_>,
         parsed_element: ParsedSyntax,
     ) -> biome_parser::parse_recovery::RecoveryResult {
-        parsed_element.or_recover(p, &SelectionListParseRecovery, expected_any_selection)
+        parsed_element.or_recover(
+            p,
+            &VariableDefinitionListParseRecovery,
+            expected_variable_definition,
+        )
     }
 }
 
+struct VariableDefinitionListParseRecovery;
+
+impl ParseRecovery for VariableDefinitionListParseRecovery {
+    type Kind = GraphqlSyntaxKind;
+    type Parser<'source> = GraphqlParser<'source>;
+    const RECOVERED_KIND: Self::Kind = GRAPHQL_VARIABLE_DEFINITION;
+
+    fn is_at_recovered(&self, p: &mut Self::Parser<'_>) -> bool {
+        is_at_variable(p)
+    }
+}
+
+/// https://spec.graphql.org/October2021/#sec-Language.Operations.Query-shorthand
+#[inline]
+pub(crate) fn parse_operation_definition(p: &mut GraphqlParser) -> ParsedSyntax {
+    if !is_at_operation(p) {
+        return Absent;
+    }
+
+    if !p.at_ts(OPERATION_TYPE) {
+        return parse_selection_set(p);
+    }
+
+    let m = p.start();
+    {
+        let m = p.start();
+        p.bump_ts(OPERATION_TYPE);
+        m.complete(p, GRAPHQL_OPERATION_TYPE);
+    }
+
+    // we don't need diagnostic here, because name is optional
+    parse_name(p).ok();
+    // we don't need diagnostic here, because variable definitions are optional
+    parse_variable_definitions(p).ok();
+
+    DirectiveList.parse_list(p);
+    parse_selection_set(p).or_add_diagnostic(p, expected_selection_set);
+
+    Present(m.complete(p, GRAPHQL_OPERATION_DEFINITION))
+}
+
+#[inline]
+fn parse_selection_set(p: &mut GraphqlParser) -> ParsedSyntax {
```

Comment:
Since we can change the `parse_selection_set` function to return a `CompleteMarker` type because the function doesn't return an `Absent` variant, we can remove unnecessary diagnostics on call sides.